### ISSUE TITLE
magic property get followed by another get returns different object when property type is not fakeable

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#312]
```csharp
public class SealedProperties
{
    public sealed class Duple
    {
        public int First { get; set; }
        public int Second { get; set; }

        public Duple(int first, int second)
        {
            First = first;
            Second = second;
        }

        public override string ToString()
        {
            return string.Concat('(', First, ", ", Second, ")");
        }
    }

    public interface InterfaceWithADuple
    {
        Duple Duple { get; set; }
    }

    [Test]
    public void Should_be_able_to_get_the_same_sealed_type_twice()
    {
        var fake = A.Fake<InterfaceWithADuple>();

        var initialDuple = fake.Duple;

        initialDuple.First = 1;
        initialDuple.Second = 2;

        var fetchedDuple = fake.Duple;

        Assert.That(fetchedDuple, Is.SameAs(initialDuple));
    }
}
```

Fails:

```
  Expected: same as <(1, 2)>
  But was:  <(0, 0)>

   at NUnit.Framework.Assert.That(Object actual, IResolveConstraint expression, String message, Object[] args)
   at NUnit.Framework.Assert.That(Object actual, IResolveConstraint expression)
   at FakeItEasyQuestions.SealedProperties.Should_be_able_to_get_the_same_sealed_type_twice() in SealedProperties.cs: line 59
```

(Really, setting the First and Second wasn't necessary. It just gives a more dramatic error message.)

I found this while working on #293 and trying to use an array as a return type. 
The cause is that `FakeManager.AutoFakePropetyRule` checks to see if the `TypeIsFakable` before saying that the rule is applicable to an incoming call. 

I'd be inclined to say that we should allow the rule if we can make a Dummy for the type, but I think that stirs up #156, which I don't mind, but it could be another month of @patrik-hagne and I going in circles. :wink:

**Comment by blairconrad** [comment]
I would like to fix this in 2.0.0.
My preference would be to fix it by 

1. having `AutoFakePropertyRule` always apply to the getter and 
2. return the a `Dummy` (the result supplied by `DefaultReturnValueRule`, currently used by unconfigured methods and getters that don't return a fakeable type).

This would

1. make sure that successive calls to the getter would return the same value  (the thrust of this issue), and
2. align the return values with unconfigured methods that return the same type (which is really the thrust of #156)

@patrik-hagne and I have had a difference of opinion on returning a Dummy to make the return value in this case. His concern is that you could then not configure a property chain such as 

```
A.CallTo(() => fakeType.Foo.Bar).Returns("some return value");
```

_when `Foo` returns a type that has an applicable `IDummyFactory`implementation_, because then the dummy factory would return (presumably) a non-fake object and its `Bar` method couldn't be configured.
I said, "Then the user gets what they ask for. If they bothered making a `DummyFactory` for the type that `Foo` returns, they probably _want_ to get the Dummy type back and in this case they should suck it up and do the less convenient configuration."

(which would look like
```
A.CallTo(() => fakeType.Foo).Returns(myFakeFooishThing);
A.CallTo(()=>myFakeFooishThing.Bar).Returns("some return value");
```
)
Then there was a long conversation about the nature of Dummies and Fakes and so on and we never came to an agreement.

However, one thing that we agree on is

> Whenever you access a property the first time the default return value is taken from the DefaultReturnValueRule, now that value is stored so that the same value can be returned the next time the same property is accessed. — _@patrik-hagne  in https://github.com/FakeItEasy/FakeItEasy/issues/156#issuecomment-28328103_

So, to satisfy _this_ issue, I think I can just change the `AutoFakePropertyRule` so it applies to all returned value types. In order to preserve as much of the existing (desirable) behaviour as we have, I would return a Fake if the type is fakeable and otherwise return a Dummy (which is what DefaultReturnValueRule returns), but _in either case_ cache the return value so the user will get the same object back again on the next call.

Objections?

**Comment by adamralph** [comment]
I tend to get quickly confused in any conversation about dummies, but after re-reading the wiki definition I think I have a better handle on what they are.

When you say "I would return a Fake if the type is fakeable and otherwise return a Dummy" are you actually just saying "I would return a Dummy"? Part of the resolution of how to create a dummy is whether the type is fakeable, no?

**Comment by blairconrad** [comment]
I'm not quite saying that. (I _want_ to say that. But this is the point over which @patrik-hagne and I  can not come to agreement.)

You are correct that part of the resolution of how to create a dummy is based on whether the type is fakeable. In many (I have no data for this, but my gut tells me by far most) cases, a Dummy for a fakeable type will be a Fake.

However, if the users have defined a `IDummyFactory` (formerly `IDummyDefinition`) that matches the type, then the Dummy will be whatever the factory creates. Might be a fake. Might be a real thing.

I think this behaviour is desirable. @patrik-hagne would prefer to see a Fake come back from the auto faked property, to support property definition chaining.

**Comment by adamralph** [comment]
I think I'm  with you on this. The proposal you are suggesting follows the principal of least astonishment and I agree with your view on property chains. Most people won't be defining dummy factories anyway, and those that do, should be advanced enough to realise that their overriding of the dummy value with a non-fake means they can't configure a method on the object.

**Comment by blairconrad** [comment]
Unfortunately, there does not seem to be a consensus on this. The impasse formerly reached in #156 remains.
However, everyone seems to agree on fixing _this_ issue, and it wouldn't preclude fixing #156 in the future, so I've sent a pull request.

**Comment by adamralph** [comment]
So we have consensus on this issue, but not on #156?

**Comment by blairconrad** [comment]
That is my understanding.

The current behaviour for unconfigured properties is that a `get` on a property with fakeable return value type will return a fake, and subsequent gets will yield the same instance.

If the return type is unfakeable, the return value will be generated using the same default return values for methods (this means it'll be a Dummy or the default value for the type). But subsequent gets will not yield the same instance (unless a `IDummyFactory` takes pains to return the same instance, I suppose).

From comments in the offline e-mail discussion the owners had, everyone agrees (or nobody objected) to caching the property get value _no matter what the return type_. 

To wit, @patrik-hagne, whom I quote without prior permission, says he prefers this workflow for unconfigured properties when a `get` is performed:

> 1. Try creating a fake.
> 1. If that fails, try to create a dummy.
> 1. If that fails, use null.
> 1. Configure the property to return the value for all future calls.
> 1. Return the value.

(I think `null` could probably be replaced with "default value for the type" in that list.) 

As I say, I believe everyone agrees on point 4 in the list - we should keep returning the same value for all future calls. This is what I'm proposing to change here.

The only point of contention that I know of is that I would prefer to strike point 1 from the list, and have the properties return the same values as unconfigured methods (a Dummy or default value for the type (note that if the type were fakeable, the Dummy would be a Fake, unless the Dummy were a `Task` or `Lazy` or the user had provided an `IDummyFactory`)). This is the crux of #156.

At worst, as I see it, falling back to creating a Dummy if we can't create a Fake, and caching the value in either case, is an improvement over the current situation.

**Comment by adamralph** [comment]
OK, sounds good to me! I'll get onto the PR ASAP.

**Comment by blairconrad** [comment]
Oh, I don't know that there's any rush. But thanks!

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
